File: hostel_management_system_backend/Controllers/AmenitiesController.cs
using Microsoft.AspNetCore.Mvc;

[ApiController]
[Route("api/[controller]")]
public sealed class AmenitiesController : ControllerBase
{
    private readonly IAmenitiesService _service;

    public AmenitiesController(IAmenitiesService service)
    {
        _service = service;
    }

    [HttpGet]
    public async Task<ActionResult<List<AmenityReadDto>>> GetAll(CancellationToken cancellationToken)
    {
        var amenities = await _service.GetAllAsync(cancellationToken);
        return Ok(amenities);
    }

    [HttpGet("{id:guid}")]
    public async Task<ActionResult<AmenityReadDto>> GetById(Guid id, CancellationToken cancellationToken)
    {
        var amenity = await _service.GetByIdAsync(id, cancellationToken);
        return amenity is null ? NotFound() : Ok(amenity);
    }

    [HttpPost]
    public async Task<ActionResult<AmenityReadDto>> Create([FromBody] AmenityCreateDto dto, CancellationToken cancellationToken)
    {
        var created = await _service.CreateAsync(dto, cancellationToken);
        return CreatedAtAction(nameof(GetById), new { id = created.Id }, created);
    }

    [HttpPut("{id:guid}")]
    public async Task<ActionResult<AmenityReadDto>> Update(Guid id, [FromBody] AmenityUpdateDto dto, CancellationToken cancellationToken)
    {
        var updated = await _service.UpdateAsync(id, dto, cancellationToken);
        return updated is null ? NotFound() : Ok(updated);
    }

    [HttpDelete("{id:guid}")]
    public async Task<IActionResult> Delete(Guid id, CancellationToken cancellationToken)
    {
        var deleted = await _service.DeleteAsync(id, cancellationToken);
        return deleted ? NoContent() : NotFound();
    }
}

File: src/amenities/amenities.controller.ts
import {
  Body,
  Controller,
  Delete,
  Get,
  HttpCode,
  HttpStatus,
  Param,
  ParseUUIDPipe,
  Post,
  Put,
} from '@nestjs/common';

import { AmenityCreateDto, AmenityReadDto, AmenityUpdateDto } from './dto/amenities.dto';
import { AmenitiesService } from './amenities.service';

@Controller('amenities')
export class AmenitiesController {
  constructor(private readonly amenitiesService: AmenitiesService) {}

  @Get()
  getAll(): Promise<AmenityReadDto[]> {
    return this.amenitiesService.getAll();
  }

  @Get(':id')
  getById(@Param('id', ParseUUIDPipe) id: string): Promise<AmenityReadDto> {
    return this.amenitiesService.getById(id);
  }

  @Post()
  create(@Body() dto: AmenityCreateDto): Promise<AmenityReadDto> {
    return this.amenitiesService.create(dto);
  }

  @Put(':id')
  update(
    @Param('id', ParseUUIDPipe) id: string,
    @Body() dto: AmenityUpdateDto,
  ): Promise<AmenityReadDto> {
    return this.amenitiesService.update(id, dto);
  }

  @Delete(':id')
  @HttpCode(HttpStatus.NO_CONTENT)
  async delete(@Param('id', ParseUUIDPipe) id: string) {
    await this.amenitiesService.delete(id);
  }
}
